### ISSUE TITLE
enable the linker to automatically choose the segment for a fragment from a list of multiple segment names

### ISSUE THREAD

**Issue by sircodalotkob** [cc65/cc65#2541]
On systems with banked or paged memory, it can be a tedious job to manually assign a large number of freely relocatable data assets to the available segments in order to fill each slot efficiently. This task could be suitable to be performed automatically by the linker.

Let's take a look at this very simple example:

```
memory
{
	bank_00:		start = $0000, size = $2000, fill = yes, fillval = $ff;
	bank_01:		start = $0000, size = $2000, fill = yes, fillval = $ff;
	bank_02:		start = $0000, size = $2000, fill = yes, fillval = $ff;
	bank_03:		start = $0000, size = $2000, fill = yes, fillval = $ff;
}

segments
{
	data_00:		load = bank_00;
	data_01:		load = bank_01;
	data_02:		load = bank_02;
	data_03:		load = bank_03;
}
```

Currently, for any fragment, no matter if source, binary or generated code, we need to assign it to a segment:

```
  .segment "data_00"
asset_00:
  .incbin "file_00"

  .segment "data_00"
asset_01:
  .incbin "file_01"

…
```

This is certainly manageable for a few assets, but for several hundred, it's not.

The idea is to give the linker an optional choice where to put a fragment. It could look like this:

```
  .segment "data_00 | data_01 | data_02 | data_03"
asset_00:
  .incbin "file_00"

  .segment "data_00 | data_01 | data_02 | data_03"
asset_01:
  .incbin "file_01"

…
```

The presence of a choice would flag the following fragment as a _free fragment_. The linker could first assign all non-free fragments like before and then place each free fragment in one of their desired segments where fits.

**Comment by mrdudz** [comment]
This should probably come with a "best effort" automatic placement algorithm too.... an interesting (non trivial) problem :)

**Comment by sircodalotkob** [comment]
Interestingly, sorting the free fragments by size, starting with the largest, and putting them in the first free of their requested slots one by one does already yield astoundingly good results.

**Comment by mrdudz** [comment]
> Interestingly, sorting the free fragments by size, starting with the largest, and putting them in the first free of their requested slots one by one does already yield astoundingly good results.

yes, thats a common approach - the interesting non trivial stuff starts when you aim for a more optimal result than that :)

**Comment by sircodalotkob** [comment]
Very true. The trivial approach would already help a lot though. Nothing to be said against an additional option to choose from a selection of multiple, possibly better fragment placement algorithms as a feature request to the feature request. :-)

**Comment by kugelfuhr** [comment]
I don't want do disappoint anyone, but to implement this feature you will have to rewrite the assembler from scratch. That's because the assembler needs to know what goes into one segment to calculate label offsets and  code sizes.

Think about zeropage/absolute addressing: The assembler needs to know how many bytes an instruction has and for this reason cannot really cope with forward references. Now think about not only one byte more or less but whole chunks of data that may be part of a segment or not. 

And it's not done with rewriting the assembler, there's additional linker support required.

So this is one of the issues that will live forever cluttering the issue list. 

Please note that I'm not judging on the usefulness of the feature described here. I'm just saying that the chance for having it implemented is as close to zero as it gets. At least in the next 20 years or so.

**Comment by mrdudz** [comment]
I am not sure this is entirely impossible - eg Jackasser solved this problem (for the "Eye of the Beholder" port") using an external script, so he didn't even have to modify the assembler (or linker). BUT it then requires multiple passes.
